Add new guide for configuring mtls between gateway & authorino/limitador (#1054)
Kuadrant/kuadrant-operator: doc/install/mtls-configuration.md

# Configure mTLS between the Gateway and Kuadrant components

## Overview

> This guide includes manual steps to enable mTLS between the gateway and Kuadrant components.

Reviewer: ```suggestion
This guide includes manual steps to enable mTLS between an Istio provided gateway and the Kuadrant components.
```

If you use an AuthPolicy or RateLimitPolicy, there will be communication between the gateway and the respective Kuadrant components at request time. This communication happens between the Wasm plugin in Envoy proxy, and Authorino or Limitador.

> At the time of writing there is [an RFC](https://github.com/Kuadrant/architecture/pull/110) discussing how to add this capability as a feature of the Kuadrant operator. If you are interested in having that feature or influencing how it is delivered, please engage on that pull request.

Reviewer: ```suggestion
At the time of writing there is [an RFC](https://github.com/Kuadrant/architecture/pull/110) discussing how to add mTLS capabilities as a feature of the Kuadrant operator. If you are interested in having that feature or influencing how it is delivered, please engage on that pull request.
```

## Enabling mTLS

### Kuadrant components

As the Kuadrant components (Authorino & Limitador) are already part of the service mesh in Istio, mTLS can be enabled after an Envoy proxy sidecar is deployed along side them.
To do this, apply the Istio sidecar label to both Deployments templates.

```bash
kubectl -n kuadrant-system patch deployment authorino \
  -p '{"spec":{"template":{"metadata":{"labels":{"sidecar.istio.io/inject":"true"}}}}}'

kubectl -n kuadrant-system patch deployment limitador-limitador \
  -p '{"spec":{"template":{"metadata":{"labels":{"sidecar.istio.io/inject":"true"}}}}}'
```

You should see the number of containers in either pod increase from 1 to 2, as the `istio-proxy` is added to the pods. This change will force all traffic to those pods to go through the proxy. However, mTLS is not enabled yet.

### Envoy Filter

The next step enables mTLS for traffic originating in the gateway (where the Wasm plugin executes), going to the Kuadrant components.

> This requires modifying the EnvoyFilters directly.

Reviewer 2: I'm actually surprised this works. In theory, any events related to the following CRs should reconcile the EnvoyFilter back to no-TLS form:
- Kuadrant
- GatewayClass
- Gateway
- HTTPRoute
- AuthPolicy
- EnvoyFilter

Author: I'll add a caveat that it may get undone, and they should allow for that by having some job or gitops thing reconcile it back.
I'll also link to the RFC where they can engage on it potential being a feature of kuadrant.

Here is an example `transport_socket` configuration to add to both the `kuadrant-auth-api-gateway` and `kuadrant-ratelimiting-api-gateway` EnvoyFilter resources.

> This should be added below `spec.configPatches[].patch.value`.

Reviewer 2: I wonder if "below" is unequivocal enough. To be more precise, I think we want to say the value of `spec.configPatches[].patch.value` (which is a YAML object) to be extended with this snippet below. I.e. for `transport_socket` to be added as a sibling of `name`, `type`, ..., `load_assignment`, is that right?

Author: Yes. I'll udpate this to be clearer

The `tls_certificate_sds_secret_configs` configures Envoy to obtain client certificates and private keys via the Secret Discovery Service (SDS) over GRPC.
The `validation_context_sds_secret_config` configures Envoy to obtain the root CA certificates via SDS (over GRPC) to validate the server's certificate.

### Istio configuration

The last step is to ensure authorino and limitador are configured to require and accept mTLS connections.
In Istio, this is done by creating a [PeerAuthentication](https://istio.io/latest/docs/reference/config/security/peer_authentication/) resource where the `mtls` mode is set to `STRICT`.
The below command will enable STRICT mode on all pods with Istio sidecar injection in the `kuadrant-system` namespace.

```bash
kubectl apply -f - <<EOF
apiVersion: security.istio.io/v1
kind: PeerAuthentication
metadata:
  name: default
  namespace: kuadrant-system
spec:
  mtls:
    mode: STRICT
EOF
```

> If you prefer to only able mTLS for a specific component, you can modify just the EnvoyFilter and Deployment for that component.

Reviewer 2: ```suggestion
If you prefer to only enable mTLS for a specific component, you can modify just the EnvoyFilter and Deployment for that component.
```
